runtime/compact.c

```diff
@@ -275,75 +221,58 @@ static void do_compaction (intnat new_allocation_policy)
 
       chend = ch + Chunk_size (ch);
       while ((char *) p < chend){
-        word q = *p;
+        header_t h = * (header_t *) p;
+        size_t sz;
 
-        if (Ecolor (q) == 0 || Tag_ehd (q) == Infix_tag){
-          /* There were (normal or infix) pointers to this block. */
-          size_t sz;
+        while (Is_gray_hd (h)) h = * dptr (h);
+        sz = Whsize_hd (h);
+
+        CAMLassert (!Is_black_hd (h));
+        CAMLassert (!Is_gray_hd (h));
+        if (Is_white_hd (h)){
+          word q;
           tag_t t;
           char *newadr;
-#ifdef WITH_PROFINFO
-          uintnat profinfo;
-#endif
-          word *infixes = NULL;
 
-          while (Ecolor (q) == 0) q = * (word *) q;
-          sz = Whsize_ehd (q);
-          t = Tag_ehd (q);
-#ifdef WITH_PROFINFO
-          profinfo = Profinfo_ehd (q);
-#endif
-          if (t == Infix_tag){
-            /* Get the original header of this block. */
-            infixes = p + sz;
-            q = *infixes;
-            CAMLassert (Ecolor (q) == 2);
-            while (Ecolor (q) != 3) q = * (word *) (q & ~(uintnat)3);
-            sz = Whsize_ehd (q);
-            t = Tag_ehd (q);
-          }
+          t = Tag_hd (h);
+          CAMLassert (t != Infix_tag);
 
           newadr = compact_allocate (Bsize_wsize (sz));
           q = *p;
-          while (Ecolor (q) == 0){
-            word next = * (word *) q;
-            * (word *) q = (word) Val_hp (newadr);
-            q = next;
+          while (Is_gray_hd (q)){
+            word *pp = dptr (q);
+            q = *pp;
+            *pp = (word) Val_hp (newadr);
           }
-          *p = Make_header_with_profinfo (Wosize_whsize (sz), t, Caml_white,
-            profinfo);
-
-          if (infixes != NULL){
-            /* Rebuild the infix headers and revert the infix pointers. */
-            while (Ecolor ((word) infixes) != 3){
-              infixes = (word *) ((word) infixes & ~(uintnat) 3);
-              q = *infixes;
-              while (Ecolor (q) == 2){
-                word next;
-                q = (word) q & ~(uintnat) 3;
-                next = * (word *) q;
-                * (word *) q = (word) Val_hp ((word *) newadr + (infixes - p));
-                q = next;
+          CAMLassert (q == h);
+          *p = q;
+
+          if (t == Closure_tag){
+            /* Revert the infix pointers. */
+            mlsize_t i, startenv, v;
+
+            v = Val_hp (p);
+            startenv = Start_env_closinfo (Closinfo_val (v));
+            i = 0;
+            while (1){
+              mlsize_t arity = Arity_closinfo (Field (v, i+1));
```

Comment:
Arities are signed (negative arity = tupled function) so `intnat` or just `int` is a better type.  (Just `int` because arities are between -128 and +127.)